engine/access/state_stream/backend/event_provider.go

```diff
@@ -0,0 +1,175 @@
+type EventsResponse struct {
+	BlockID          flow.Identifier
+	Height           uint64
+	Events           flow.EventsList
+	BlockTimestamp   time.Time
+	ExecutorMetadata access.ExecutorMetadata
+}
+
+// EventsProvider retrieves events by block height. It can be configured to retrieve events from
+// the events indexer(if available) or using a dedicated callback to query it from other sources.
+type EventsProvider struct {
+	log              zerolog.Logger
+	headers          storage.Headers
+	execDataProvider ExecutionDataProvider
+
+	fetchFromLocalCache bool
+	execResultProvider  optimistic_sync.ExecutionResultInfoProvider
+	execStateCache      optimistic_sync.ExecutionStateCache
+}
+
+func NewEventsProvider(
+	log zerolog.Logger,
+	headers storage.Headers,
+	execDataProvider ExecutionDataProvider,
+	fetchFromLocalCache bool,
+	execResultProvider optimistic_sync.ExecutionResultInfoProvider,
+	execStateCache optimistic_sync.ExecutionStateCache,
+) *EventsProvider {
+	return &EventsProvider{
+		log:                 log,
+		headers:             headers,
+		execDataProvider:    execDataProvider,
+		fetchFromLocalCache: fetchFromLocalCache,
+		execResultProvider:  execResultProvider,
+		execStateCache:      execStateCache,
+	}
+}
+
+// Events return events for a given block height.
+//
+// Expected error returns during normal operations:
+//   - [codes.NotFound]: If a block header for the specified block height is not found.
+func (b *EventsProvider) Events(
+	ctx context.Context,
+	height uint64,
+	criteria optimistic_sync.Criteria,
+) (*EventsResponse, error) {
+	var response *EventsResponse
+	var err error
+
+	if b.fetchFromLocalCache {
+		response, err = b.getEventsFromStorage(height, criteria)
+	} else {
+		response, err = b.getEventsFromExecutionData(ctx, height, criteria)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get events: %w", err)
+	}
+
+	header, err := b.headers.ByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get header for height %d: %w", height, err)
+	}
+	response.BlockTimestamp = time.UnixMilli(int64(header.Timestamp)).UTC()
+
+	if b.log.GetLevel() == zerolog.TraceLevel {
+		b.log.Trace().
+			Hex("block_id", logging.ID(response.BlockID)).
+			Uint64("height", height).
+			Int("events", len(response.Events)).
+			Msg("sending events")
+	}
+
+	return response, nil
+}
+
+// getEventsFromExecutionData returns the events for a given height extract from the execution data.
+// Expected errors:
+// - error: An error indicating issues with getting execution data for block
+func (b *EventsProvider) getEventsFromExecutionData(
+	ctx context.Context,
+	height uint64,
+	criteria optimistic_sync.Criteria,
+) (*EventsResponse, error) {
+	blockID, err := b.headers.BlockIDByHeight(height)
+	if err != nil {
+		return nil, fmt.Errorf("could not get header for height %d: %w", height, err)
+	}
+
+	result, err := b.execResultProvider.ExecutionResultInfo(
+		blockID,
+		criteria,
+	)
```

Comment:
nit: don't break the line up for a small number of arguments. this just reduces the amount of code you can see at once which makes the code less readable imo. only break split arguments onto separate lines if there are a lot (e.g. 5+), or the line is long (e.g. 120+ character)
```suggestion
	result, err := b.execResultProvider.ExecutionResultInfo(blockID, criteria)
```